xilem_web/src/memoized_await.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2024 the Xilem Authors and the Druid Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{future::Future, marker::PhantomData};
+
+use wasm_bindgen::{closure::Closure, JsCast, UnwrapThrowExt};
+use wasm_bindgen_futures::spawn_local;
+use xilem_core::{MessageResult, Mut, NoElement, View, ViewId, ViewPathTracker};
+
+use crate::{DynMessage, OptionalAction, ViewCtx};
+
+/// Await a future returned by `init_future` invoked with the argument `data`, `callback` is called with the output of the future. `init_future` will be invoked again, when `data` changes. Use [`memoized_await`] for construction of this [`View`]
+pub struct MemoizedAwait<State, Action, OA, InitFuture, Data, Callback, F, FOut> {
+    init_future: InitFuture,
+    data: Data,
+    callback: Callback,
+    debounce: usize,
+    reset_debounce_on_update: bool,
+    #[allow(clippy::type_complexity)]
+    phantom: PhantomData<fn() -> (State, Action, OA, F, FOut)>,
+}
+
+impl<State, Action, OA, InitFuture, Data, CB, F, FOut>
+    MemoizedAwait<State, Action, OA, InitFuture, Data, CB, F, FOut>
+{
+    /// Debounce the `init_future` function, when `data` updates,
+    /// when `reset_debounce_on_update == false` then this throttles updates each `millisecconds`
+    pub fn debounce(mut self, milliseconds: usize) -> Self {
+        self.debounce = milliseconds;
+        self
+    }
+
+    /// When `reset` is `true`, everytime `data` updates, the debounce timeout is cleared until `init_future` is invoked.
+    /// This is only effective when `debounce > 0`
+    pub fn reset_debounce_on_update(mut self, reset: bool) -> Self {
```

Comment:
Thinking about it, mabye this should be: 
```suggestion
    pub fn reset_debounce_timout_on_update(mut self, reset: bool) -> Self {
```